test/e2e/gateway_api_test.go

```diff
@@ -0,0 +1,189 @@
+//go:build e2e
+// +build e2e
+
+package e2e
+
+import (
+	"context"
+	"testing"
+	"time"
+
+	configv1 "github.com/openshift/api/config/v1"
+	gwapi "github.com/openshift/cluster-ingress-operator/pkg/operator/controller/gatewayapi"
+
+	corev1 "k8s.io/api/core/v1"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/wait"
+
+	"sigs.k8s.io/gateway-api/apis/v1beta1"
+)
+
+const (
+	testHostname = "*.example.com"
+)
+
+// Define here so we can delete them in the defer section
+var gateway v1beta1.Gateway
+var gatewayClass v1beta1.GatewayClass
+
+// TestGatewayAPIResources tests that basic functions for Gateway API Custom Resources are functional.
+// It specifically verifies that when the GatewayAPI feature gate is enabled, then a user can
+// create a GatewayClass, Gateway, and HTTPRoute.
+func TestGatewayAPIResources(t *testing.T) {
+	t.Parallel()
+
+	// Get the cluster feature gate
+	featureGate := &configv1.FeatureGate{}
+	err := wait.PollImmediate(1*time.Second, 1*time.Minute, func() (bool, error) {
+		name := types.NamespacedName{"default", "cluster"}
+		if err := kclient.Get(context.TODO(), name, featureGate); err != nil {
+			t.Logf("failed to get feature gate %s: %v", gwapi.FeatureGateName, err)
+			return false, nil
+		}
+		return true, nil
+	})
+	if err != nil {
+		t.Fatalf("failed to find feature gate: %v", err)
+	}
+
+	// Create a test namespace if needed
+	ns := buildNamespace("test-namespace")
+	if err := kclient.Create(context.TODO(), ns); err != nil {
+		if !errors.IsAlreadyExists(err) {
+			t.Fatalf("error creating namespace: %v", err)
+		}
+	}
+	defer func() {
+		// Deleting the namespace also deletes the httproute
+		if err := kclient.Delete(context.TODO(), ns); err != nil {
+			t.Fatalf("failed to delete test namespace %s: %v", ns.Name, err)
+		}
+	}()
+
+	// Check if the feature gate is disabled, and if it is, make sure no GWAPI CRDs can be created
+	if !gwapi.FeatureIsEnabled(gwapi.FeatureGateName, featureGate) {
+		// First make sure nothing can happen with the Gateway API CRDs when the feature is disabled
+		gatewayClassFailed, gatewayFailed, httpRouteFailed := testCrds(t, ns.Name, false)
+		if !(gatewayClassFailed && gatewayFailed && httpRouteFailed) {
+			t.Fatalf("feature gate was not enabled, but gateway api objects could be created")
```

Comment:
Would it make sense to delete whatever did get created in this case?  Although you might want to be careful to distinguish between resources that were created (clean these up) and those that were already there (maybe don't clean those up).

Reply:
The defer function will take care of deleting everything in any case.